### PR TITLE
Capture observed dependence on unknowns in simplified nonlinear systems

### PR DESCRIPTION
**Pull request by hersle**
This fixes the incorrect Jacobian in #2858.

## Checklist

- [x] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API

### CODE REVIEW

**Comment by hersle** [comment]
Would it be better if, for example,
```julia
using ModelingToolkit
@variables x y z
@parameters σ ρ β
@mtkbuild ns = NonlinearSystem([
    0 ~ σ * (y - x)
    0 ~ x * (ρ - z) - y
    0 ~ x * y - β * z
])
J = calculate_jacobian(ns)
```
did not expand observed derivatives like
```
2×2 Matrix{Num}:
 (-1 - z + ρ)*σ        -x*σ
    2x*(-z + ρ)  -β - (x^2)
```
but rather kept them in a form like
```
2×2 Matrix{Num}:
 (-1 + Differential(x)(y))*σ     Differential(z)(y)*σ
    y + x*Differential(x)(y)  x*Differential(z)(y) -β
```
?

File: src/systems/nonlinear/nonlinearsystem.jl [hunk 1]
```diff
@@ -193,8 +193,13 @@ function calculate_jacobian(sys::NonlinearSystem; sparse = false, simplify = fal
         return cache[1]
     end
 
-    rhs = [eq.rhs for eq in equations(sys)]
+    # observed equations may depend on unknowns, so substitute them in first
+    # TODO: must do the same fix in e.g. calculate_hessian?
+    # TODO: rather keep observed derivatives unexpanded, like "Differential(obs)(expr)"?
+    obs = map(eq -> eq.lhs => eq.rhs, observed(sys))
+    rhs = map(eq -> substitute(eq.rhs, obs), equations(sys))
```
Comment (ChrisRackauckas) [inline src/systems/nonlinear/nonlinearsystem.jl:200]: This won't catch a chain of observed, it would need a fixed point sub I think

Comment (hersle) [reply src/systems/nonlinear/nonlinearsystem.jl]: Like in the most recent commit? Can you think of a simple example on to test?

Comment (ChrisRackauckas) [reply src/systems/nonlinear/nonlinearsystem.jl]: `z ~ y, `y ~ x`, `x^2 - 2z` should hit it, since if you just plug in z you get y.

Comment (hersle) [reply src/systems/nonlinear/nonlinearsystem.jl]: Not quite after structural simplification, it seems.
But I added `0 ~ x^2 + 2*z + y`, `z ~ y`, `y ~ x`, which seems to do the trick.

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**